test/test_eth_registrar_controller.js

```diff
@@ -58,150 +61,192 @@ contract('ETHRegistrarController', function (accounts) {
+	});
+
+	it('should include the owner in the commitment', async () => {
+		await controller.commit(await controller.makeCommitment("newname2", accounts[2], secret));
+
+		await advanceTime((await controller.minCommitmentAge()).toNumber());
+		var balanceBefore = await web3.eth.getBalance(controller.address);
+		await expectFailure(controller.register("newname2", registrantAccount, 28 * DAYS, secret, {value: 28 * DAYS, gasPrice: 0}));
+	});
+
+	it('should reject duplicate registrations', async () => {
+		await controller.commit(await controller.makeCommitment("newname", registrantAccount, secret));
+
+		await advanceTime((await controller.minCommitmentAge()).toNumber());
+		var balanceBefore = await web3.eth.getBalance(controller.address);
+		await expectFailure(controller.register("newname", registrantAccount, 28 * DAYS, secret, {value: 28 * DAYS, gasPrice: 0}));
+	});
+
+	it('should reject for expired commitments', async () => {
+		await controller.commit(await controller.makeCommitment("newname2", registrantAccount, secret));
+
+		await advanceTime((await controller.maxCommitmentAge()).toNumber() + 1);
+		var balanceBefore = await web3.eth.getBalance(controller.address);
+		await expectFailure(controller.register("newname2", registrantAccount, 28 * DAYS, secret, {value: 28 * DAYS, gasPrice: 0}));
+	});
+
+	it('should allow anyone to renew a name', async () => {
+		var expires = await baseRegistrar.nameExpires(sha3("newname"));
+		var balanceBefore = await web3.eth.getBalance(controller.address);
+		await controller.renew("newname", 86400, {value: 86400 + 1});
+		var newExpires = await baseRegistrar.nameExpires(sha3("newname"));
+		assert.equal(newExpires.toNumber() - expires.toNumber(), 86400);
+		assert.equal((await web3.eth.getBalance(controller.address)) - balanceBefore, 86400);
+	});
+
+	it('should require sufficient value for a renewal', async () => {
+		await expectFailure(controller.renew("newname", 86400));
+	});
+
+	it('should permit bulk renewal', async () => {
+		var balanceBefore = await web3.eth.getBalance(controller.address);
+		const nameExpires = await baseRegistrar.nameExpires(sha3("newname"));
+		await controller.renewAll(["newname", "newname"], 86400, NULL_ADDRESS, {value: 86400 * 2 + 1});
+		assert.equal((await baseRegistrar.nameExpires(sha3("newname"))) - nameExpires, 86400 * 2);
+		assert.equal((await web3.eth.getBalance(controller.address)) - balanceBefore, 86400 * 2);
+	});
+
+	it('should require sufficient value for a bulk renewal', async () => {
+		await expectFailure(controller.renewAll(["name", "newname"], 86400, NULL_ADDRESS, {value: 86401}));
+	});
+
+	it('should allow the registrar owner to withdraw funds', async () => {
+		await controller.withdraw({gasPrice: 0, from: ownerAccount});
+		assert.equal(await web3.eth.getBalance(controller.address), 0);
+	});
+
+	it('should permit registration with a valid referrer', async () => {
+		await controller.setReferralFee(100);
```

Comment:
Would be nice to have a separate test to make sure that random person cannot set the referral fee